File: openmetadata-airflow-apis/openmetadata_managed_apis/workflows/ingestion/auto_classification.py
#  Copyright 2021 Collate
#  Licensed under the Apache License, Version 2.0 (the "License");
#  you may not use this file except in compliance with the License.
#  You may obtain a copy of the License at
#  http://www.apache.org/licenses/LICENSE-2.0
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
"""
Auto Classification DAG function builder
"""
import json

from airflow import DAG
from openmetadata_managed_apis.utils.logger import set_operator_logger
from openmetadata_managed_apis.workflows.ingestion.common import build_dag, build_source

from metadata.generated.schema.entity.services.ingestionPipelines.ingestionPipeline import (
    IngestionPipeline,
)
from metadata.generated.schema.metadataIngestion.workflow import (
    LogLevels,
    OpenMetadataWorkflowConfig,
    Processor,
    Sink,
    WorkflowConfig,
)
from metadata.workflow.classification import AutoClassificationWorkflow


def auto_classification_workflow(workflow_config: OpenMetadataWorkflowConfig):
    """
    Task that creates and runs the auto classification workflow.

    The workflow_config gets cooked form the incoming
    ingestionPipeline.

    This is the callable used to create the PythonOperator
    """

    set_operator_logger(workflow_config)

    config = json.loads(workflow_config.model_dump_json(exclude_defaults=False))
    workflow = AutoClassificationWorkflow.create(config)

    workflow.execute()
    workflow.raise_from_status()
    workflow.print_status()
    workflow.stop()


def build_auto_classification_workflow_config(
    ingestion_pipeline: IngestionPipeline,
) -> OpenMetadataWorkflowConfig:
    """
    Given an airflow_pipeline, prepare the workflow config JSON
    """
    workflow_config = OpenMetadataWorkflowConfig(
        source=build_source(ingestion_pipeline),
        sink=Sink(
            type="metadata-rest",
            config={},
        ),
        processor=Processor(
            type="orm-profiler",
            config={},
        ),
        workflowConfig=WorkflowConfig(
            loggerLevel=ingestion_pipeline.loggerLevel or LogLevels.INFO,
            openMetadataServerConfig=ingestion_pipeline.openMetadataServerConnection,
        ),
        ingestionPipelineFQN=ingestion_pipeline.fullyQualifiedName.root,
    )

    return workflow_config


def build_auto_classification_dag(ingestion_pipeline: IngestionPipeline) -> DAG:
    """
    Build a simple metadata workflow DAG
    """
    workflow_config = build_auto_classification_workflow_config(ingestion_pipeline)
    dag = build_dag(
        task_name="auto_classification_task",
        ingestion_pipeline=ingestion_pipeline,
        workflow_config=workflow_config,
        workflow_fn=auto_classification_workflow,
    )

    return dag

File: openmetadata-ui/src/main/resources/ui/playwright/utils/autoClassification.ts
/*
 *  Copyright 2024 Collate.
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *  http://www.apache.org/licenses/LICENSE-2.0
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { Page } from '@playwright/test';
import MysqlIngestionClass from '../support/entity/ingestion/MySqlIngestionClass';
import { getApiContext, toastNotification } from './common';
import { visitServiceDetailsPage } from './service';
import { waitForAllLoadersToDisappear } from './entity';

export const addAndTriggerAutoClassificationPipeline = async (
  page: Page,
  mysqlService: MysqlIngestionClass
) => {
  const { apiContext } = await getApiContext(page);

  await visitServiceDetailsPage(
    page,
    {
      type: mysqlService.category,
      name: mysqlService.name,
      displayName: mysqlService.name,
    },
    true
  );

  // Add auto classification ingestion
  await page.click('[data-testid="agents"]');

  const metadataTab = page.locator('[data-testid="metadata-sub-tab"]');
  if (await metadataTab.isVisible()) {
    await metadataTab.click();
  }

  await page.click('[data-testid="add-new-ingestion-button"]');

  await page
    .locator('.ant-dropdown:visible [data-menu-id*="autoClassification"]')
    .waitFor();

  await page.locator('[data-menu-id*="autoClassification"]').waitFor();

  await page.click('[data-menu-id*="autoClassification"]');

  // Fill the auto classification form details
  await page.locator('#root\\/tableFilterPattern\\/includes').waitFor();

  await mysqlService.fillIngestionDetails(page);

  await page.click('[data-testid="submit-btn"]');

  // Make sure we create ingestion with None schedule to avoid conflict between Airflow and Argo behavior
  await mysqlService.scheduleIngestion(page);

  await page.click('[data-testid="view-service-button"]');

  // Header available once page loads
  await waitForAllLoadersToDisappear(page);
  await page.getByTestId('agents').click();
  const metadataTab2 = page.locator('[data-testid="metadata-sub-tab"]');
  if (await metadataTab2.isVisible()) {
    await metadataTab2.click();
  }
  await page
    .getByLabel('agents')
    .getByTestId('loader')
    .waitFor({ state: 'detached' });

  const response = await apiContext
    .get(
      `/api/v1/services/ingestionPipelines?service=${encodeURIComponent(
        mysqlService.name
      )}&pipelineType=autoClassification&serviceType=databaseService&limit=1`
    )
    .then((res) => res.json());

  // eslint-disable-next-line playwright/no-wait-for-timeout -- pipeline deployment settling time
  await page.waitForTimeout(3000);

  await page.click(
    `[data-row-key*="${response.data[0].name}"] [data-testid="more-actions"]`
  );
  await page.getByTestId('run-button').click();

  await toastNotification(page, `Pipeline triggered successfully!`);

  // eslint-disable-next-line playwright/no-wait-for-timeout -- wait for latest pipeline run results
  await page.waitForTimeout(2000);

  await mysqlService.handleIngestionRetry('autoClassification', page);
};
